Implement experimental basic job execution progress reporting hook

nulltoken opened the pull request:
## Pull request description
Fix #147

The idea behind this proposal is to provide a **non stable** reporting mechanism.
While allowing me to scratch my own #147 itch and potentially receiving potential feedback/feature request from other early bleeding edge adopters, this should also allow us to slowly evolve this toward the requirements of #36 of an more feature rich observer.

### PR meta checklist
- [x] Pull request is targeted at `main` branch for code   
- [x] Pull request is linked to all related issues, if any.

### Code PR specific checklist
- [x] My code follows the code style of this project and AspNetCore coding guidelines.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
- [x] I have updated the appropriate sub section in the _CHANGELOG.md_.
- [x] I have added, updated or removed tests to according to my changes.
  - [x] All tests passed.

nulltoken commented on src/NCronJob/Execution/JobExecutionProgressMonitor.cs:
```diff
@@ -0,0 +1,150 @@
+#pragma warning restore S101 // Types should be named in PascalCase
+#pragma warning restore CA1707 // Identifiers should not contain underscores
+{
+    internal ExecutionProgress_Unstable(JobRun run)
```
``xxx_unstable` is a standard pattern in the nodejs ecosystem when you want to release early a work in progress (or an experimental feature) but want to be free to introduce any breaking change outside of a major major version release.

Basically, it's a "Use at your own risk, may change or even disappear without prior notice" warning label.

linkdotnet replied:
To be fair, the `ExperimentalAttribute` would be enough for me. That signals already what is going on.

nulltoken replied:
Fixed

nulltoken commented on Directory.Build.props:
```diff
@@ -16,6 +16,8 @@
         <AnalysisLevel>latest</AnalysisLevel>
         <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
+
+        <NoWarn>$(NoWarn);NCRONJOB_REPORTING</NoWarn>
```
I haven't been able to successfully move that silencing to `.editorconfig` :-/

cf. https://learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/experimental-overview#suppress-warnings

linkdotnet commented:
I took a similar approach some time ago. Also @falvarez1 started working on this.
I can remember that I tried to use Mediator-Pattern for this.

Especially because if a job doesn't get scheduled or executed you also need some hook. (For example: Multiple jobs are scheduled at the same time, but only one can be executed).

As far as I can tell, only somewhat "successful" runs will be registered in the current implementation. 

We might also want to remove `JobState` as we have now two things somewhat reflecting the same state.

nulltoken commented:
> Especially because if a job doesn't get scheduled or executed you also need some hook. (For example: Multiple jobs are scheduled at the same time, but only one can be executed).

@linkdotnet The current approach is hooked on the JobRun state machine. However, when no JobRun is created, nothing is indeed published.

>As far as I can tell, only somewhat "successful" runs will be registered in the current implementation.

That's not the intent. Every state change _should_ be reported.

In one of the latest pushes, I've added a test that showcases an Expired JobRun (as part of this, I discovered that Expired wasn't considered as a final state and fixed it).

nulltoken commented on src/NCronJob/NCronJob.csproj:
```diff
@@ -37,7 +37,16 @@
+      Documentation will be eventually hand rolled before merge.
+
+      -->
+		<GenerateDocumentationFile>false</GenerateDocumentationFile>
```
Savagely hacked on purpose. To be reverted before merge

linkdotnet replied:
Interesting. For what did you need the this?

nulltoken replied:
Because I'm (shamelessly) more interested in getting feedback from the CI server test runs on different platforms than documenting the public types.

All the "Missing XML comment for publicly visible type or member xxxx" error messages were making my "Developer Experience" a lot less rich :wink:

linkdotnet replied:
I am not sure how to properly handle that.

Opting out might lead to missing XML annotations. It is very strict (well at least not if you use DEBUG locally, but that doesn't help in your case)

nulltoken replied:
> Savagely hacked on purpose. To be reverted before merge

Once you think the API is ok. I'll write the xml documentation and revert this change. Before the merge.

nulltoken replied:
Reverted as documentation has been added.

nulltoken commented:
> We might also want to remove `JobState` as we have now two things somewhat reflecting the same state.

If that ok with you I'd rather tackle that in a follow up PR. I attempted to do that, but in the end this generated a lot of diff noise that made this proposal harder to read.

Besides, there's a entry `JobState.Crashed` that doesn't seem to be leveraged anywhere, and I wasn't sure what to do with it 🤔

nulltoken commented on src/NCronJob/Scheduler/JobProcessor.cs:
```diff
@@ -33,7 +33,10 @@ public async Task ProcessJobAsync(JobRun jobRun, CancellationToken cancellationT
             await jobExecutor.RunJob(jobRun, cancellationToken).ConfigureAwait(false);
 
-            jobRun.NotifyStateChange(JobStateType.Completed);
+            if (!jobRun.IsCompleted)
```
Without this, a faulted job (for instance) would eventually be marked as Completed.

nulltoken commented on src/NCronJob/Registry/JobRun.cs:
```diff
@@ -88,16 +140,21 @@ public void AddState(JobState state)
     public void NotifyStateChange(JobStateType type, string message = "")
     {
-        if (CurrentState.Type == type)
+        if (CurrentState.Type == type || IsCompleted)
```
Without this, a faulted job (for instance) would eventually be marked as Completed.

nulltoken commented:
> Another idea I toyed around (and maybe you had similar thoughts) is to have a mediator in the middle and do all the things just by messaging, removing a bit of the coupling. Ultimately I didn't go through as it seemed a lot of noise when I spiked something together and the overall amount of objects would be higher than for example your current approach.

As a lot of things happens between all the types, I was willing to disrupt the current code in the least possible way in order to prevent any regression. Hence, why I tiptoed and end up plugging in the JobRun state machine.

The trajectory would be to cover the current proposal with as many tests as possible in order to ensure it works as expected, and merge it as is.

The tests would provide us with a safety net would you want to revamp the internal messaging architecture.

nulltoken commented:
Just added some additional test coverage with regards to StartupJobs.

@linkdotnet Do you see any other existing use cases you'd like to see covered?

nulltoken commented on .github/workflows/dotnet.yml:
```diff
@@ -17,6 +17,7 @@ jobs:
   build:
 
     runs-on: ${{ matrix.os }}
+    timeout-minutes: 10
```
Had a couple of builds that went stuck for hours. 

![image](https://github.com/user-attachments/assets/4d53f152-2150-4361-8a1d-5de07ce56c0f)


At first, I thought it was a concurrency related issues, but it wasn't eventually the case. Maybe a transient GitHub actions issue?

Whatever, let's try to bail out early when that happens and avoid wasting those precious free build minutes.

linkdotnet replied:
I like that!

linkdotnet commented on src/NCronJob/Observer/ExecutionProgress.cs:
```diff
@@ -0,0 +1,40 @@
+using System.Diagnostics.CodeAnalysis;
+
+namespace NCronJob.Experimental;
```
I would also remove the namespace here. Not really discoverable and the attribute is enough (IMHO)

nulltoken replied:
Fixed

linkdotnet commented on src/NCronJob/Observer/ExecutionState.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace NCronJob.Experimental;
+
+[Experimental("NCRONJOB_OBSERVER")]
```
Same same

linkdotnet replied:
Love it - thanks!

linkdotnet commented:
> Just added some additional test coverage with regards to StartupJobs.
> 
> @linkdotnet Do you see any other existing use cases you'd like to see covered?

You mean besides Startup Jobs?

One case would be, that jobs are **not** executed. Right now, we are using just timeouts for that, which isn't ideal to be honest (in many cases). If we can "see" that a job isn't scheduled for a reason, i think that would make some tests more verbose and easier to understand.

linkdotnet commented on src/NCronJob/Registry/JobRun.cs:
```diff
@@ -1,36 +1,50 @@
         var jobRunId = Guid.NewGuid();
 
         JobRunId = jobRunId;
+        ParentJobRunId = parentJob is not null ? parentJob.JobRunId : jobRunId;
```
That is interesting. Is there a reason not to set `ParanetJobRunId` to `null` if there isn't any parent?

nulltoken commented:
>One case would be, that jobs are not executed. Right now, we are using just timeouts for that, which isn't ideal to be honest (in many cases). If we can "see" that a job isn't scheduled for a reason, i think that would make some tests more verbose and easier to understand.

@linkdotnet Can you please orient me to some pieces of code relevant to that (in the lib and in the tests)?

nulltoken commented on src/NCronJob/Scheduler/JobQueueManager.cs:
```diff
@@ -47,6 +47,8 @@ public void RemoveQueue(string queueName)
         {
             if (jobQueues.TryRemove(queueName, out var jobQueue))
             {
+                // TODO: Should we first cancel the potentially existing JobRuns here?
```
@linkdotnet Thoughts?

linkdotnet replied:
That is a very good question.
I wouldn't cancel already running jobs.

linkdotnet replied:
**BUT** that is just my experience.

nulltoken replied:
> I wouldn't cancel already running jobs.

That would make sense.

I _think_ the queue may also contain scheduled pending jobs which I believe should be cancelled before being cleared out. I'll work on this.

nulltoken replied:
> I _think_ the queue may also contain scheduled pending jobs which I believe should be cancelled before being cleared out. I'll work on this.

@linkdotnet Done

nulltoken replied:
>> I think the queue may also contain scheduled pending jobs which I believe should be cancelled before being cleared out. I'll work on this.

>@linkdotnet Done

Actually, I rolled it back. It uncovers a hell of its own. Logged as #161

linkdotnet commented:
> > One case would be, that jobs are not executed. Right now, we are using just timeouts for that, which isn't ideal to be honest (in many cases). If we can "see" that a job isn't scheduled for a reason, i think that would make some tests more verbose and easier to understand.
> 
> @linkdotnet Can you please orient me to some pieces of code relevant to that (in the lib and in the tests)?

For example: https://github.com/NCronJob-Dev/NCronJob/blob/main/src/NCronJob/JobExecutionContext.cs#L44

Skipping dependent Jobs. Right now, we are using measure like this:
```csharp
(await WaitForJobsOrTimeout(1, TimeSpan.FromMilliseconds(500))).ShouldBe(false);
```
<sup>See: https://github.com/NCronJob-Dev/NCronJob/blob/main/tests/NCronJob.Tests/RunDependentJobTests.cs#L74</sup>

linkdotnet commented:
@nulltoken On an unrelated side note, do you have any GitHub Sponsor page, Kofi or so?

nulltoken commented:
> For example: https://github.com/NCronJob-Dev/NCronJob/blob/main/src/NCronJob/JobExecutionContext.cs#L44
> 
> Skipping dependent Jobs. Right now, we are using measure like this:

This one is a whole different beast. JobRuns of dependent jobs aren't even created yet when the skipping is done. 
Logged as #160

nulltoken commented:
@linkdotnet I think I'm done from the code standpoint on this first step and would welcome a review. 

It fulfills the need of #147 and reports the start/completion of an orchestration (whether it would be made of simple job or a root job along with its dependencies).

It also contains the following fixes (covered by the tests):
- Consider an `Expired` job state as a final state
- Prevents the further state changes of job already in its final state

Identified additional features have been logged in separate features:
- #160
- #161

linkdotnet commented:
> @linkdotnet I think I'm done from the code standpoint on this first step and would welcome a review.
> 
> It fulfills the need of #147 and reports the start/completion of an orchestration (whether it would be made of simple job or a root job along with its dependencies).
> 
> It also contains the following fixes (covered by the tests):
> 
> * Consider an `Expired` job state as a final state
> * Prevents the further state changes of job already in its final state
> 
> Identified additional features have been logged in separate features:
> 
> * [Report on skipped jobs #160](https://github.com/NCronJob-Dev/NCronJob/issues/160)
> * [Cancel scheduled (but not started yet) jobs on IRuntimeRegistry.Remove() #161](https://github.com/NCronJob-Dev/NCronJob/issues/161)

First of all: Thanks very much for the PR! Really nice. 

I guess it was the best decision to track those two things independently and not tackle them in this PR. 

That said - let's bring this in. You might wanna add this to the `CHANGELOG.md` and *maybe* add some sentences and use cases to the documentation.

nulltoken commented:
> > You might wanna add this to the `CHANGELOG.md` and _maybe_ add some sentences and use cases to the documentation.
> 
> 👍 Will work on that and report when it's done.

@linkdotnet Done

linkdotnet commented on docs/advanced/observing-job-progress.md:
```diff
@@ -0,0 +1,223 @@
+        var reporter = app.Services.GetRequiredService<IJobExecutionProgressReporter>();
+
+        // ...enlist a new subscriber to it...
+        IDisposable subscriber = reporter.Register(Subscriber);
```
Maybe `using var subscriber` would be more idiomatic for .NET devs

nulltoken replied:
I'd prefer keeping the `IDisposable` type under the light. That's a rather important aspect of the design.

linkdotnet replied:
Can you elaborate a bit on this?

Handling the `Dispose` explicitly can lead to many pitfalls. 

Especially what if,during `Register` and `Dispose` there is an exception? The example should be somewhat idiomatic and lead developers to a pit of success.

nulltoken replied:
>Can you elaborate a bit on this?

This example isn't the final way this should be implemented. It is contrived.

In the real life, there are few chances the subscriber will be a private function. 
Chances are it'll rather be a type built and injected in the DI container. In that design, the type itself should be disposable (and taking care of disposing the registered subscribers) and the disposal would be taken care by the DI container.

However, as I was willing to build a not too lengthy example, I went with this.

While working on #36, it's even possible we'll have to build something similar (or an abstraction of it) and make it available to the users so that they don't have to reinvent the wheel.

>Especially what if,during Register and Dispose there is an exception? The example should be somewhat idiomatic and lead developers to a pit of success.

This is an experimental feature. It's not finished. It may not work and/or change. It comes with rough corners and sharp edges :wink:

The final developer experience will be ready when this feature is deemed stable.

The only goal of that sample is to showcase how the pieces move together and give a chance to potential early adopter to "play" with it.

It's even described as such in the previous paragraph (*"Below a very simple approach...."*).

If you prefer, I can emphasize this with some words putting under the light this isn't a "copy-and-paste-production-ready" code.

linkdotnet replied:
Don't get me wrong, this is totally fine to merge.  That is why I already approved.

> This is an experimental feature. It's not finished. It may not work and/or change. It comes with rough corners and sharp edges 

True - but for me, it was more about how the user code is written. That is totally independent of our stability with the feature. 

The code handles the disposable in a suboptimal way - but I get your verbosity here.

nulltoken replied:
>True - but for me, it was more about how the user code is written. That is totally independent of our stability with the feature.

>The code handles the disposable in a suboptimal way

I agree and we'll refine later that part of the documentation 

>- but I get your verbosity here.

Thanks for coping with me :wink:

> Don't get me wrong, this is totally fine to merge. That is why I already approved.

Let's get this in then!

linkdotnet reviewed (approved):
LGTM
